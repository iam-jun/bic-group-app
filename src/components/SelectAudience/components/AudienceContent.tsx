import {
  StyleProp, StyleSheet, View, ViewStyle,
} from 'react-native';
import React, { FC, useEffect, useState } from 'react';
import { ExtendedTheme, useTheme } from '@react-navigation/native';
import { Avatar, Button } from '~/baseComponents';
import { IGroup } from '~/interfaces/IGroup';
import { GroupPrivacyDetail } from '~/constants/privacyTypes';
import { IconType } from '~/resources/icons';
import Text from '~/baseComponents/Text';
import { spacing } from '~/theme';
import Tag from '~/baseComponents/Tag';
import Icon from '~/baseComponents/Icon';

export interface GroupItemProps {
  style?: StyleProp<ViewStyle>;
  item: IGroup;
  nameLines?: number;

  onPress: (item: IGroup, isChecked: boolean) => void
  shouldBeChecked: (item: IGroup) => boolean;
  isSingleSelect?: boolean;
}

const AudienceContent: FC<GroupItemProps> = ({
  style,
  item,
  nameLines = 1,

  onPress,
  shouldBeChecked,
  isSingleSelect,
}) => {
  const theme: ExtendedTheme = useTheme();
  const styles = themeStyles();
  const { colors } = theme;
  const [checked, setChecked] = useState(shouldBeChecked(item));

  const {
    icon, name, privacy, community,
  } = item || {};
  const { name: communityName } = community || {};
  const privacyIcon = GroupPrivacyDetail[privacy]?.icon as IconType;

  useEffect(() => {
    setChecked(shouldBeChecked(item));
  }, [shouldBeChecked(item)]);

  const _onPress = () => {
    const newValue = !checked;
    onPress(item, newValue);
    setChecked(newValue);
  };

  const checkBoxStyles = {
    unselect: {
      iconName: !isSingleSelect && 'Square' as IconType,
      iconColor: !isSingleSelect && colors.neutral20,
    },
    selected: {
      iconName: isSingleSelect ? 'Check' : 'SquareCheckSolid' as IconType,
      iconColor: isSingleSelect ? colors.green50 : colors.blue50,
    },
  };

  const { iconName, iconColor } = checked ? checkBoxStyles.selected : checkBoxStyles.unselect;

  return (
    <Button
      testID="group_item.container"
      style={[styles.container, style]}
      onPress={_onPress}
    >
      <Avatar.Base
        source={icon}
        privacyIcon={privacyIcon}
      />
      <View style={styles.textContainer}>
        <Text.BodyM
          style={styles.textName}
          color={theme.colors.neutral60}
          numberOfLines={nameLines}
        >
          {name}
        </Text.BodyM>
        <Tag
          textProps={{ numberOfLines: 1 }}
          style={styles.tagContainer}
          type="secondary"
          label={communityName}
          disabled
        />
      </View>
      <View>
        {
          !!iconName && <Icon icon={iconName} size={22} tintColor={iconColor} />
        }
      </View>
    </Button>
  );
};

const themeStyles = () => StyleSheet.create({
  container: {
    flex: 1,
    flexDirection: 'row',
    alignItems: 'center',
  },
  textContainer: {
    flex: 1,
    justifyContent: 'center',
    paddingHorizontal: spacing.padding.base,
  },
  row: {
    flexDirection: 'row',
    alignItems: 'center',
  },
  textName: {
    width: '100%',
    justifyContent: 'center',
  },
  tagContainer: {
    alignSelf: 'baseline',
    marginTop: spacing.margin.xTiny,
  },
});

export default AudienceContent;
