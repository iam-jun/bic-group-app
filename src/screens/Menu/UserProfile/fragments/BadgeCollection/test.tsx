import React from 'react';
import {
  View, Dimensions, FlatList, StyleSheet,
} from 'react-native';
import { ExtendedTheme, useTheme } from '@react-navigation/native';
import Avatar from '~/baseComponents/Avatar';
import Icon from '~/baseComponents/Icon';
import spacing, { borderRadius } from '~/theme/spacing';

const Grid = ({ data }: {data: any[]}) => {
  const deviceWidth = Dimensions.get('window').width;

  const itemWidth = 48; // Constant item width
  const numColumns = Math.floor((deviceWidth - 10 * 2) / (itemWidth + 6 * 2));
  const rowWidth = numColumns * (itemWidth + 6 * 2);
  const paddingLeft = (deviceWidth - (10 * 2) - rowWidth) / 2;

  const theme: ExtendedTheme = useTheme();
  const { colors } = theme;
  const styles = themeStyles(theme);

  const isSelected = false;

  const renderGridItem = ({ item, index }: any) =>
    // Render each grid item here
    (
      <View
        key={index}
        style={{
          margin: 6,
          borderWidth: isSelected ? 2 : 1,
          borderColor: isSelected ? colors.purple50 : colors.neutral5,
          borderRadius: borderRadius.pill,
        }}
      >
        <Avatar.Medium
          isRounded
          source={{ uri: item?.iconUrl }}
        />
        {isSelected
          ? (
            <View style={styles.iconChose}>
              <Icon size={10} icon="Check" tintColor={colors.white} />
            </View>
          )
          : null}
      </View>
    );

  return (
    <View style={{
      flex: 1,
      justifyContent: 'center',
      alignItems: 'center',
      paddingHorizontal: 10,
      paddingBottom: 10,
    }}
    >

      <View style={{
        // paddingLeft,
      }}
      >
        <FlatList
          data={data}
          renderItem={renderGridItem}
          keyExtractor={(item, index) => index.toString()}
          numColumns={numColumns}
        />
      </View>
    </View>
  );
};

const themeStyles = (theme: ExtendedTheme) => {
  const { colors } = theme;

  return StyleSheet.create({
    container: {
      flex: 1,
      paddingTop: spacing.padding.large,
    },
    iconChose: {
      backgroundColor: colors.purple50,
      borderRadius: borderRadius.pill,
      height: 20,
      width: 20,
      alignItems: 'center',
      justifyContent: 'center',
      position: 'absolute',
      zIndex: 1,
      right: 0,
    },
  });
};

export default Grid;
