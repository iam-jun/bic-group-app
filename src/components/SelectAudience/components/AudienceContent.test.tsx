import React from 'react';
import { fireEvent, renderWithRedux } from '~/test/testUtils';
import AudienceContent from './AudienceContent';
import { mockResponseFlatAudiences } from '~/test/mock_data/audiences';

describe('AudienceContent component', () => {
  it('should be pressable', () => {
    const shouldBeChecked = jest.fn();
    const onPress = jest.fn();
    const itemData = mockResponseFlatAudiences.data[0];

    const component = renderWithRedux(
      <AudienceContent
        item={itemData as any}
        onPress={onPress}
        shouldBeChecked={shouldBeChecked}
      />,
    );

    expect(shouldBeChecked).toBeCalled();

    const itemAudience = component.getByTestId('group_item.container');

    fireEvent.press(itemAudience);

    expect(onPress).toBeCalled();
  });

  it('should be check', () => {
    const shouldBeChecked = jest.fn().mockReturnValue(true);
    const onPress = jest.fn();
    const itemData = mockResponseFlatAudiences.data[0];

    const component = renderWithRedux(
      <AudienceContent
        item={itemData as any}
        onPress={onPress}
        shouldBeChecked={shouldBeChecked}
      />,
    );

    expect(component).toMatchSnapshot();
  });

  it('should be uncheck', () => {
    const shouldBeChecked = jest.fn().mockReturnValue(false);
    const onPress = jest.fn();
    const itemData = mockResponseFlatAudiences.data[0];

    const component = renderWithRedux(
      <AudienceContent
        item={itemData as any}
        onPress={onPress}
        shouldBeChecked={shouldBeChecked}
      />,
    );

    expect(component).toMatchSnapshot();
  });
});
