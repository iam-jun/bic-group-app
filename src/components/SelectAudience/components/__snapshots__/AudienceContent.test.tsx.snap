// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AudienceContent component should be check 1`] = `
<View
  accessible={true}
  collapsable={false}
  focusable={true}
  onClick={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    Object {
      "alignItems": "center",
      "flex": 1,
      "flexDirection": "row",
      "opacity": 1,
    }
  }
  testID="group_item.container"
>
  <View
    style={
      Array [
        Object {
          "borderRadius": 6,
          "height": 40,
          "width": 40,
        },
        undefined,
      ]
    }
    testID="avatar"
  >
    <View
      style={
        Array [
          Object {
            "borderRadius": 6,
            "height": 40,
            "width": 40,
          },
          Object {},
        ]
      }
      testID="avatar_container"
    >
      <View
        style={
          Array [
            Object {
              "overflow": "hidden",
            },
            Array [
              Object {
                "borderRadius": 6,
                "height": 40,
                "width": 40,
              },
              Object {
                "borderColor": "#F2F5F8",
                "borderWidth": 2,
              },
              Object {
                "backgroundColor": "#FFFFFF",
              },
            ],
          ]
        }
      >
        <FastImageView
          onFastImageError={[Function]}
          resizeMode="cover"
          source={
            Object {
              "uri": "https://bic-dev-entity-attribute-s3-bucket.s3.ap-southeast-1.amazonaws.com/static/group/default-avatar.png",
            }
          }
          style={
            Object {
              "bottom": 0,
              "left": 0,
              "position": "absolute",
              "right": 0,
              "top": 0,
            }
          }
          testID="avatar.image"
        />
      </View>
      <View
        style={
          Object {
            "alignItems": "center",
            "backgroundColor": "#007AFF",
            "borderRadius": 200,
            "bottom": -2,
            "height": 20,
            "justifyContent": "center",
            "position": "absolute",
            "right": -2,
            "width": 20,
          }
        }
      >
        <View
          accessibilityState={
            Object {
              "disabled": true,
            }
          }
          accessible={true}
          collapsable={false}
          focusable={false}
          hitSlop={
            Object {
              "bottom": 10,
              "left": 10,
              "right": 10,
              "top": 10,
            }
          }
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            Object {
              "opacity": 1,
            }
          }
        >
          <View
            style={
              Array [
                Object {
                  "alignItems": "center",
                  "flexDirection": "row",
                },
                undefined,
                Object {
                  "backgroundColor": undefined,
                },
              ]
            }
          >
            <View
              style={
                Array [
                  undefined,
                  undefined,
                  undefined,
                ]
              }
            >
              <View />
            </View>
          </View>
        </View>
      </View>
    </View>
  </View>
  <View
    style={
      Object {
        "flex": 1,
        "justifyContent": "center",
        "paddingHorizontal": 12,
      }
    }
  >
    <Text
      numberOfLines={1}
      style={
        Array [
          Object {
            "color": "#212745",
            "fontFamily": "BeVietnamPro-Light",
            "fontSize": 14,
          },
          Object {
            "color": "#282E53",
          },
          Object {
            "justifyContent": "center",
            "width": "100%",
          },
        ]
      }
    >
      Right Thinking
    </Text>
    <View
      accessibilityState={
        Object {
          "disabled": true,
        }
      }
      accessible={true}
      collapsable={false}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "alignItems": "center",
          "alignSelf": "baseline",
          "backgroundColor": "#EBF4FF",
          "borderRadius": 100,
          "flexDirection": "row",
          "height": 24,
          "justifyContent": "center",
          "marginRight": 8,
          "marginTop": 2,
          "opacity": 1,
          "paddingLeft": 4,
          "paddingRight": 8,
        }
      }
      testID="tag.item"
    >
      <Text
        numberOfLines={1}
        style={
          Array [
            Object {
              "color": "#212745",
              "fontFamily": "BeVietnamPro-Light",
              "fontSize": 11,
            },
            Object {},
            Array [
              Object {
                "color": "#007AFF",
                "paddingLeft": 4,
              },
              undefined,
            ],
          ]
        }
        testID="tag.label"
      >
        Right Thinking
      </Text>
    </View>
  </View>
  <View>
    <View
      accessibilityState={
        Object {
          "disabled": true,
        }
      }
      accessible={true}
      collapsable={false}
      focusable={false}
      hitSlop={
        Object {
          "bottom": 10,
          "left": 10,
          "right": 10,
          "top": 10,
        }
      }
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "opacity": 1,
        }
      }
    >
      <View
        style={
          Array [
            Object {
              "alignItems": "center",
              "flexDirection": "row",
            },
            undefined,
            Object {
              "backgroundColor": undefined,
            },
          ]
        }
      >
        <View
          style={
            Array [
              undefined,
              undefined,
              undefined,
            ]
          }
        >
          <View />
        </View>
      </View>
    </View>
  </View>
</View>
`;

exports[`AudienceContent component should be uncheck 1`] = `
<View
  accessible={true}
  collapsable={false}
  focusable={true}
  onClick={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    Object {
      "alignItems": "center",
      "flex": 1,
      "flexDirection": "row",
      "opacity": 1,
    }
  }
  testID="group_item.container"
>
  <View
    style={
      Array [
        Object {
          "borderRadius": 6,
          "height": 40,
          "width": 40,
        },
        undefined,
      ]
    }
    testID="avatar"
  >
    <View
      style={
        Array [
          Object {
            "borderRadius": 6,
            "height": 40,
            "width": 40,
          },
          Object {},
        ]
      }
      testID="avatar_container"
    >
      <View
        style={
          Array [
            Object {
              "overflow": "hidden",
            },
            Array [
              Object {
                "borderRadius": 6,
                "height": 40,
                "width": 40,
              },
              Object {
                "borderColor": "#F2F5F8",
                "borderWidth": 2,
              },
              Object {
                "backgroundColor": "#FFFFFF",
              },
            ],
          ]
        }
      >
        <FastImageView
          onFastImageError={[Function]}
          resizeMode="cover"
          source={
            Object {
              "uri": "https://bic-dev-entity-attribute-s3-bucket.s3.ap-southeast-1.amazonaws.com/static/group/default-avatar.png",
            }
          }
          style={
            Object {
              "bottom": 0,
              "left": 0,
              "position": "absolute",
              "right": 0,
              "top": 0,
            }
          }
          testID="avatar.image"
        />
      </View>
      <View
        style={
          Object {
            "alignItems": "center",
            "backgroundColor": "#007AFF",
            "borderRadius": 200,
            "bottom": -2,
            "height": 20,
            "justifyContent": "center",
            "position": "absolute",
            "right": -2,
            "width": 20,
          }
        }
      >
        <View
          accessibilityState={
            Object {
              "disabled": true,
            }
          }
          accessible={true}
          collapsable={false}
          focusable={false}
          hitSlop={
            Object {
              "bottom": 10,
              "left": 10,
              "right": 10,
              "top": 10,
            }
          }
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            Object {
              "opacity": 1,
            }
          }
        >
          <View
            style={
              Array [
                Object {
                  "alignItems": "center",
                  "flexDirection": "row",
                },
                undefined,
                Object {
                  "backgroundColor": undefined,
                },
              ]
            }
          >
            <View
              style={
                Array [
                  undefined,
                  undefined,
                  undefined,
                ]
              }
            >
              <View />
            </View>
          </View>
        </View>
      </View>
    </View>
  </View>
  <View
    style={
      Object {
        "flex": 1,
        "justifyContent": "center",
        "paddingHorizontal": 12,
      }
    }
  >
    <Text
      numberOfLines={1}
      style={
        Array [
          Object {
            "color": "#212745",
            "fontFamily": "BeVietnamPro-Light",
            "fontSize": 14,
          },
          Object {
            "color": "#282E53",
          },
          Object {
            "justifyContent": "center",
            "width": "100%",
          },
        ]
      }
    >
      Right Thinking
    </Text>
    <View
      accessibilityState={
        Object {
          "disabled": true,
        }
      }
      accessible={true}
      collapsable={false}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "alignItems": "center",
          "alignSelf": "baseline",
          "backgroundColor": "#EBF4FF",
          "borderRadius": 100,
          "flexDirection": "row",
          "height": 24,
          "justifyContent": "center",
          "marginRight": 8,
          "marginTop": 2,
          "opacity": 1,
          "paddingLeft": 4,
          "paddingRight": 8,
        }
      }
      testID="tag.item"
    >
      <Text
        numberOfLines={1}
        style={
          Array [
            Object {
              "color": "#212745",
              "fontFamily": "BeVietnamPro-Light",
              "fontSize": 11,
            },
            Object {},
            Array [
              Object {
                "color": "#007AFF",
                "paddingLeft": 4,
              },
              undefined,
            ],
          ]
        }
        testID="tag.label"
      >
        Right Thinking
      </Text>
    </View>
  </View>
  <View>
    <View
      accessibilityState={
        Object {
          "disabled": true,
        }
      }
      accessible={true}
      collapsable={false}
      focusable={false}
      hitSlop={
        Object {
          "bottom": 10,
          "left": 10,
          "right": 10,
          "top": 10,
        }
      }
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "opacity": 1,
        }
      }
    >
      <View
        style={
          Array [
            Object {
              "alignItems": "center",
              "flexDirection": "row",
            },
            undefined,
            Object {
              "backgroundColor": undefined,
            },
          ]
        }
      >
        <View
          style={
            Array [
              undefined,
              undefined,
              undefined,
            ]
          }
        >
          <View />
        </View>
      </View>
    </View>
  </View>
</View>
`;
